pkg/tso/keyspace_group_manager.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2023 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tso
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/log"
+	mcsutils "github.com/tikv/pd/pkg/mcs/utils"
+	"github.com/tikv/pd/pkg/member"
+	"github.com/tikv/pd/pkg/storage/endpoint"
+	"github.com/tikv/pd/pkg/storage/kv"
+	"github.com/tikv/pd/pkg/utils/memberutil"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+)
+
+const (
+	// maxKeyspaceGroupCount is the max count of keyspace groups.
+	// Keyspace group in tso is the sharding unit, i.e., by the definition here,
+	// the max count of the shards we support is maxKeyspaceGroupCount. We use
+	// five-digits number (%05d) to render the keyspace group id in the storage
+	// path, so theoretically the max count is 99999 which is far beyond what's
+	// actually needed.
+	maxKeyspaceGroupCount = uint32(4096)
```

Comment:
Do we need to set the upper bound?

Reply:
ok, explicitly defined the upper bound and added the sanity check.